src/main/scala/beam/agentsim/agents/household/HouseholdCAVScheduling.scala

```diff
@@ -1,195 +1,398 @@
+    import scala.collection.mutable.{ListBuffer => MList}
+    val emptyFleetSchedule = MList.empty[CAVSchedule]
+    cavVehicles.foldLeft(householdRequests.requests.head) {
+      case (req, cav) =>
+        emptyFleetSchedule.prepend(
+          new CAVSchedule(
+            List[MobilityServiceRequest](
+              new MobilityServiceRequest(
+                None,
+                req.activity,
+                req.time,
+                req.trip,
+                req.defaultMode,
+                Init,
+                req.time
+              )
+            ),
+            cav,
+            0
+          )
+        )
+        (req)
+    }
+    // compute all the feasible schedules through
+    val feasibleSchedules = MList[CAVFleetSchedule](new CAVFleetSchedule(emptyFleetSchedule.toList, householdRequests))
+    for (request  <- householdRequests.requests;
+         schedule <- feasibleSchedules) yield {
+      val (newSchedule, feasible) = schedule.check(request, skim)
+      feasibleSchedules.prependAll(newSchedule)
+      if (!feasible) feasibleSchedules -= schedule
+    }
+    feasibleSchedules.toList
+  }
+
+  // ***
+  // get k lowest scored schedules
+  def getKBestSchedules(k: Int): List[CAVFleetSchedule] = {
+    return getAllFeasibleSchedules().sortBy(_.householdTrips.totalTravelTime).take(k)
+  }
+
+  // ***
+  def getBestScheduleWithTheLongestCAVChain(): CAVFleetSchedule = {
+    val maprank =
+      getAllFeasibleSchedules().map(x => x -> x.cavFleetSchedule.foldLeft(0)((a, b) => a + b.schedule.size)).toMap
+    val maxrank = maprank.maxBy(_._2)._2
+    maprank.filter(_._2 == maxrank).keys.toList.sortBy(_.householdTrips.totalTravelTime).take(1)(0)
+  }
+
+  // ***
+  case class CAVFleetSchedule(cavFleetSchedule: List[CAVSchedule], householdTrips: HouseholdTrips) {
 
-    def check(request: MobilityServiceRequest): Option[CAVSchedule] = {
-      val travelTime = skim(schedule.last.activity.getCoord)(request.activity.getCoord)
-      val arrivalTime = schedule.last.time + schedule.last.deltaTime + travelTime
-      val newDeltaTime = arrivalTime - request.time
-      val newCost = cost + newDeltaTime
-      var newCavSchedule: Option[CAVSchedule] = None
-      request.tag match {
-        case MSRPickup =>
-          if (occupancy == 0 && newDeltaTime < -1 * timeWindow) {
-            val relocationRequest =
-              new MobilityServiceRequest(None, request.activity, request.time - 1, newDeltaTime, MSRRelocation)
-            val newRequest = new MobilityServiceRequest(request.person, request.activity, request.time, 0, MSRPickup)
-            newCavSchedule = Some(
-              new CAVSchedule(schedule :+ relocationRequest :+ newRequest, cav, newCost, occupancy + 1)
+    // ***
+    def check(
+      request: MobilityServiceRequest,
+      skim: Map[BeamMode, Map[Coord, Map[Coord, Double]]]
+    ): (List[CAVFleetSchedule], Boolean) = {
+      import scala.collection.mutable.{ListBuffer => MList}
```

Comment:
Can you change this to MutableList so that it is more explicit

Reply:
Yes, but my point was to make the M more explicit, so the suggestion would change to MutableListBuffer